File: backend/app/schemas/websocket.py
"""WebSocket event schemas for real-time job progress updates.

This module defines the event types and payload schemas for WebSocket
communication between the backend and frontend clients.

Event Types:
- job_started: Fired when a job begins processing
- progress_update: Fired periodically with progress percentage
- job_completed: Fired when a job finishes successfully
- job_failed: Fired when a job fails with error details
- connection_established: Acknowledgment of successful connection
- error: General error event
"""

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field

from app.models.generation_job import JobStatus


class WebSocketEventType(str, Enum):
    """WebSocket event types for job progress updates."""

    # Connection events
    CONNECTION_ESTABLISHED = "connection_established"
    CONNECTION_ERROR = "connection_error"

    # Job lifecycle events
    JOB_STARTED = "job_started"
    PROGRESS_UPDATE = "progress_update"
    JOB_COMPLETED = "job_completed"
    JOB_FAILED = "job_failed"
    JOB_CANCELLED = "job_cancelled"

    # Client commands
    PING = "ping"
    PONG = "pong"


class BaseWebSocketEvent(BaseModel):
    """Base class for all WebSocket events."""

    type: WebSocketEventType = Field(..., description="Event type identifier")
    timestamp: datetime = Field(
        default_factory=datetime.utcnow, description="Event timestamp (UTC)"
    )


class ConnectionEstablishedEvent(BaseWebSocketEvent):
    """Event sent when WebSocket connection is established successfully."""

    type: WebSocketEventType = WebSocketEventType.CONNECTION_ESTABLISHED
    job_id: int = Field(..., description="Job ID being monitored")
    current_status: JobStatus = Field(..., description="Current job status")
    current_progress: float = Field(..., ge=0.0, le=100.0, description="Current progress")
    message: str = Field(default="Connected to job progress stream")


class ConnectionErrorEvent(BaseWebSocketEvent):
    """Event sent when there's a connection-level error."""

    type: WebSocketEventType = WebSocketEventType.CONNECTION_ERROR
    error_code: str = Field(..., description="Error code for client handling")
    message: str = Field(..., description="Human-readable error message")


class JobStartedEvent(BaseWebSocketEvent):
    """Event sent when a job begins processing."""

    type: WebSocketEventType = WebSocketEventType.JOB_STARTED
    job_id: int = Field(..., description="Job ID")
    started_at: datetime = Field(..., description="Job start timestamp")
    agent_name: Optional[str] = Field(None, description="First agent to process")


class ProgressUpdateEvent(BaseWebSocketEvent):
    """Event sent periodically with job progress updates."""

    type: WebSocketEventType = WebSocketEventType.PROGRESS_UPDATE
    job_id: int = Field(..., description="Job ID")
    progress_pct: float = Field(..., ge=0.0, le=100.0, description="Progress percentage")
    progress_message: Optional[str] = Field(None, description="Human-readable progress status")
    agent_name: Optional[str] = Field(None, description="Current agent processing")
    stage: Optional[str] = Field(None, description="Current processing stage")


class JobCompletedEvent(BaseWebSocketEvent):
    """Event sent when a job completes successfully."""

    type: WebSocketEventType = WebSocketEventType.JOB_COMPLETED
    job_id: int = Field(..., description="Job ID")
    map_id: Optional[int] = Field(None, description="Generated map ID")
    completed_at: datetime = Field(..., description="Completion timestamp")
    total_duration_seconds: Optional[float] = Field(
        None, description="Total processing time in seconds"
    )


class JobFailedEvent(BaseWebSocketEvent):
    """Event sent when a job fails."""

    type: WebSocketEventType = WebSocketEventType.JOB_FAILED
    job_id: int = Field(..., description="Job ID")
    error_message: str = Field(..., description="Error message")
    error_code: Optional[str] = Field(None, description="Error classification code")
    can_retry: bool = Field(..., description="Whether the job can be retried")
    retry_count: int = Field(..., description="Number of retries attempted")
    max_retries: int = Field(..., description="Maximum retries allowed")
    failed_at: datetime = Field(..., description="Failure timestamp")


class JobCancelledEvent(BaseWebSocketEvent):
    """Event sent when a job is cancelled."""

    type: WebSocketEventType = WebSocketEventType.JOB_CANCELLED
    job_id: int = Field(..., description="Job ID")
    cancelled_at: datetime = Field(..., description="Cancellation timestamp")


class PingEvent(BaseWebSocketEvent):
    """Ping event for connection keep-alive."""

    type: WebSocketEventType = WebSocketEventType.PING


class PongEvent(BaseWebSocketEvent):
    """Pong response to ping for connection keep-alive."""

    type: WebSocketEventType = WebSocketEventType.PONG


# Type alias for all possible WebSocket events
WebSocketEvent = (
    ConnectionEstablishedEvent
    | ConnectionErrorEvent
    | JobStartedEvent
    | ProgressUpdateEvent
    | JobCompletedEvent
    | JobFailedEvent
    | JobCancelledEvent
    | PingEvent
    | PongEvent
)


class WebSocketMessage(BaseModel):
    """Wrapper for WebSocket messages with optional metadata."""

    event: dict = Field(..., description="Event payload")
    correlation_id: Optional[str] = Field(
        None, description="Optional correlation ID for tracking"
    )

    @classmethod
    def from_event(
        cls, event: BaseWebSocketEvent, correlation_id: Optional[str] = None
    ) -> "WebSocketMessage":
        """Create a message from an event."""
        return cls(
            event=event.model_dump(mode="json"),
            correlation_id=correlation_id,
        )

File: frontend/src/types/websocket.ts
/**
 * WebSocket Event Types and Interfaces
 *
 * TypeScript types for WebSocket communication with the backend.
 * These types mirror the backend Pydantic schemas.
 */

/**
 * Job status enum matching backend JobStatus
 */
export type JobStatus =
  | 'pending'
  | 'processing'
  | 'completed'
  | 'failed'
  | 'cancelled';

/**
 * WebSocket event types
 */
export type WebSocketEventType =
  | 'connection_established'
  | 'connection_error'
  | 'job_started'
  | 'progress_update'
  | 'job_completed'
  | 'job_failed'
  | 'job_cancelled'
  | 'ping'
  | 'pong';

/**
 * Base event interface
 */
export interface BaseWebSocketEvent {
  type: WebSocketEventType;
  timestamp: string;
}

/**
 * Connection established event
 */
export interface ConnectionEstablishedEvent extends BaseWebSocketEvent {
  type: 'connection_established';
  job_id: number;
  current_status: JobStatus;
  current_progress: number;
  message: string;
}

/**
 * Connection error event
 */
export interface ConnectionErrorEvent extends BaseWebSocketEvent {
  type: 'connection_error';
  error_code: string;
  message: string;
}

/**
 * Job started event
 */
export interface JobStartedEvent extends BaseWebSocketEvent {
  type: 'job_started';
  job_id: number;
  started_at: string;
  agent_name?: string;
}

/**
 * Progress update event
 */
export interface ProgressUpdateEvent extends BaseWebSocketEvent {
  type: 'progress_update';
  job_id: number;
  progress_pct: number;
  progress_message?: string;
  agent_name?: string;
  stage?: string;
}

/**
 * Job completed event
 */
export interface JobCompletedEvent extends BaseWebSocketEvent {
  type: 'job_completed';
  job_id: number;
  map_id?: number;
  completed_at: string;
  total_duration_seconds?: number;
}

/**
 * Job failed event
 */
export interface JobFailedEvent extends BaseWebSocketEvent {
  type: 'job_failed';
  job_id: number;
  error_message: string;
  error_code?: string;
  can_retry: boolean;
  retry_count: number;
  max_retries: number;
  failed_at: string;
}

/**
 * Job cancelled event
 */
export interface JobCancelledEvent extends BaseWebSocketEvent {
  type: 'job_cancelled';
  job_id: number;
  cancelled_at: string;
}

/**
 * Ping event (sent by client)
 */
export interface PingEvent extends BaseWebSocketEvent {
  type: 'ping';
}

/**
 * Pong event (received from server)
 */
export interface PongEvent extends BaseWebSocketEvent {
  type: 'pong';
}

/**
 * Union type for all WebSocket events
 */
export type WebSocketEvent =
  | ConnectionEstablishedEvent
  | ConnectionErrorEvent
  | JobStartedEvent
  | ProgressUpdateEvent
  | JobCompletedEvent
  | JobFailedEvent
  | JobCancelledEvent
  | PingEvent
  | PongEvent;

/**
 * WebSocket message wrapper
 */
export interface WebSocketMessage {
  event: WebSocketEvent;
  correlation_id?: string;
}

/**
 * WebSocket connection state
 */
export type WebSocketConnectionState =
  | 'connecting'
  | 'connected'
  | 'disconnected'
  | 'reconnecting'
  | 'error';

/**
 * Job progress state tracked by the hook
 */
export interface JobProgressState {
  jobId: number;
  status: JobStatus;
  progressPct: number;
  progressMessage?: string;
  agentName?: string;
  stage?: string;
  startedAt?: string;
  completedAt?: string;
  errorMessage?: string;
  errorCode?: string;
  canRetry: boolean;
  mapId?: number;
}

/**
 * Options for useWebSocket hook
 */
export interface UseWebSocketOptions {
  /** Auto-reconnect on disconnect (default: true) */
  autoReconnect?: boolean;
  /** Maximum reconnection attempts (default: 5) */
  maxReconnectAttempts?: number;
  /** Base delay for exponential backoff in ms (default: 1000) */
  reconnectBaseDelay?: number;
  /** Maximum delay between reconnection attempts in ms (default: 30000) */
  reconnectMaxDelay?: number;
  /** Ping interval in ms for keep-alive (default: 30000) */
  pingInterval?: number;
  /** Callback when connection established */
  onConnected?: (event: ConnectionEstablishedEvent) => void;
  /** Callback when job starts */
  onJobStarted?: (event: JobStartedEvent) => void;
  /** Callback when progress updates */
  onProgress?: (event: ProgressUpdateEvent) => void;
  /** Callback when job completes */
  onCompleted?: (event: JobCompletedEvent) => void;
  /** Callback when job fails */
  onFailed?: (event: JobFailedEvent) => void;
  /** Callback when connection error occurs */
  onError?: (event: ConnectionErrorEvent) => void;
}

/**
 * Return type for useWebSocket hook
 */
export interface UseWebSocketResult {
  /** Current connection state */
  connectionState: WebSocketConnectionState;
  /** Current job progress state */
  progress: JobProgressState | null;
  /** Whether currently connected */
  isConnected: boolean;
  /** Connect to job progress stream */
  connect: () => void;
  /** Disconnect from job progress stream */
  disconnect: () => void;
  /** Reconnect to job progress stream */
  reconnect: () => void;
  /** Last error message */
  error: string | null;
  /** Number of reconnection attempts */
  reconnectAttempts: number;
}
